### PR TITLE
Fixes `initFirstItem` bypass when `ui.isAccessAllowed` is defined

### PR DESCRIPTION
**Pull request by dcousens**
This pull request rewrites the `@keystone-6/auth` middleware and `isAccessAllowed` composition.

Fixes
- https://github.com/keystonejs/keystone/issues/7970
- https://github.com/keystonejs/keystone/issues/8085

Related
- https://github.com/keystonejs/keystone/issues/7662

### CODE REVIEW

File: packages/auth/src/index.ts [hunk 1]
```diff
@@ -255,44 +210,93 @@ export function createAuth<ListTypeInfo extends BaseListTypeInfo>({
...
         isAccessAllowed: async (context: KeystoneContext) => {
-          // Allow access to the adminMeta data from the /init path to correctly render that page
-          // even if the user isn't logged in (which should always be the case if they're seeing /init)
-          const headers = context.req?.headers;
-          const host = headers ? headers['x-forwarded-host'] || headers['host'] : null;
-          const url = headers?.referer ? new URL(headers.referer) : undefined;
-          const accessingInitPage =
-            url?.pathname === '/init' &&
-            url?.host === host &&
-            (await context.sudo().query[listKey].count({})) === 0;
```
Comment (dcousens) [inline packages/auth/src/index.ts:289]: I replaced this with
```
if (pathname === '/__nextjs_original-stack-frame') return
```

And for the meta endpoint the replacement for `isAccessAllowed` is simply to return `true` for this when `hasInitFirstItemConditions`.

File: packages/core/src/admin-ui/context.tsx [hunk 2]
```diff
@@ -113,13 +113,6 @@ export const useKeystone = (): {
     throw new Error('useKeystone must be called inside a KeystoneProvider component');
   }
   if (value.adminMeta.state === 'error') {
-    // If we get an "Access denied" error, it probably means the user doesn't have access to the
-    // adminMeta but has navigated (probably client-side) to a page that requires it. We reload
-    // the page so the server-side access control can run which should bounce them to the right
-    // place (or display the no-access page)
-    if (value.adminMeta.error.message === 'Access denied') {
-      window.location.reload();
-    }
```
Comment (dcousens) [inline packages/core/src/admin-ui/context.tsx:122]: We might want to think about how to do this differently, the infinite loop that can happen with this isn't helpful

File: packages/auth/src/index.ts [hunk 3]
```diff
@@ -255,44 +210,93 @@ export function createAuth<ListTypeInfo extends BaseListTypeInfo>({
...
-          return (
-            accessingInitPage ||
-            (keystoneConfig.ui?.isAccessAllowed
-              ? keystoneConfig.ui.isAccessAllowed(context)
-              : context.session !== undefined)
-          );
+          if (await hasInitFirstItemConditions(context)) return true;
```
Comment (dcousens) [inline packages/auth/src/index.ts:298]: With effectively being a bypass, I wanted the function to be isolated

**Comment by dcousens** [comment]
This will probably need to go out as a patch, independent of the upcoming work

File: packages/auth/src/index.ts (hunk 1 above)
Comment (dcousens) [reply packages/auth/src/index.ts:289]: Now replaced with https://github.com/keystonejs/keystone/blob/f6fb299cb987a27ce9c0885620dc2b68a6186ef8/packages/core/src/lib/server/createAdminUIMiddleware.ts#L46

File: packages/core/src/lib/server/createAdminUIMiddleware.ts [hunk 4]
```diff
@@ -24,37 +24,47 @@ export async function getNextApp(dev: boolean, projectAdminPath: string): Promis
...
   context: KeystoneContext,
   nextApp: NextApp
 ) {
   const handle = nextApp.getRequestHandler();
 
-  const { ui, session } = config;
-  const publicPages = ui?.publicPages ?? [];
+  const {
+    ui: { isAccessAllowed = defaultIsAccessAllowed, pageMiddleware, publicPages = [] } = {},
+  } = config;
```
Comment (dcousens) [inline packages/core/src/lib/server/createAdminUIMiddleware.ts:41]: Defaults > branching everywhere

**Review by flexdinesh — APPROVED** [review]

**Pull request merged**